### DISCUSSION TITLE
Dynamic Power Limiter außerhalb gegebenen Parametern

### DISCUSSION THREAD

**Discussion by l3ztum** [hoylabs/OpenDTU-OnBattery#212] [Q&A]
Hi, ich wollte erstmal eine Diskussion öffnen um evtl. Fehlparametrierung auszuschließen, bevor ich ein Issue auf mache. 

Ich bin gerade dabei meine Balkonanlage mit Akku aufzubauen. Habe aktuell den Akku schon da und wollte den Dynamic Power Limiter von in Version 2023.04.29 ohne angeschlossenes Solarpanel testen, was meinem aktuellen techn. und elektrischem Verständniss nach funktionieren sollte.

**Aufbau**:
* 48V Akku auf Channel0 an einem HM-1500
* Shelly 3EM mit N-Leiter über HTTP angebunden, da ich unabhängig vom MQTT Server sein möchte
* OpenDTU-OnBattery auf einem ESP32-Wroom-32 mit angeschlossenem nRF24L01+
* Verbindungen werden alle als i.O. angezeigt

Grundsätzlich wird Akkuspannung, Netzinfo etc. korrekt vom DTU erkannt. Habe dafür die Einstellungen im Screenshot
 angelegt. 48V Akku hängt auf Inverter id 0 und Channel 0. Wenn ich die  Einstellungen speichere, geht der Inverterausgang ruckzuck auf 800W und es wird ins Netz eingespeist, da der aktuelle Verbrauch nur bei ca 350W liegt.  Hat einer ein Tipp wieso hier so übersteuert wird obwohl Limit bei 200W eingestellt ist?

![Screenshot](https://user-images.githubusercontent.com/26739228/235351860-b14fb9cf-6a0a-480f-8513-e4bca1947ecc.png)

**Comment by helgeerbe** [comment]
Kannst du mal die ausgaben aus der Konsole schicken? Da müssten ja ein paar Angaben zu finden sein.
Hat du nur einen Channel oder alle 4 an der Batterie hängen?

**Comment by l3ztum** [comment]
[LogOutput.txt](https://github.com/helgeerbe/OpenDTU-OnBattery/files/11361399/LogOutput.txt)
Aktuell ist nur ein Channel an der Batterie, da ich nur bis 200W testen wollte. 
@helgeerbe die Consolen-Outputs scheinen sich mit dem Verbrauch zu decken, allerdings sieht man in der Weboberfläche, dass die tatsächlich eingestellten Werte weit über die Werte der Console hinaus gehen. Shelly Oberfläche zeigt dementsprechend dann auch eine Einspeisung an. Da hier Werte eingestellt werden, die an den Grenzen des Wechselrichters und der Batterie liegen, möchte ich den Aufbau auch nicht allzu lange aktiv lassen um Defekte zu vermeiden.

Edit: hatte jetzt testweise nochmal einen anderen Akku mit 36V dran hängen

**Reply by l3ztum** [reply]
Manuell via Weboberfläche eingestellte Werte werden im übrigen korrekt übernommen.

**Comment by helgeerbe** [comment]
Das sieht eigentlich gut aus.

````log
15:25:12.950 > [PowerLimiterClass::loop] dcVoltage: 38.00 Voltage Start Threshold: 36.00 Voltage Stop Threshold: 32.00 inverter->isProducing(): 0
15:25:12.950 > [PowerLimiterClass::loop] newPowerLimit: 168
15:25:12.950 > [PowerLimiterClass::loop] Starting up inverter...
15:25:12.950 > [PowerLimiterClass::loop] Limit Non-Persistent: 168 W
````

Ein Channel an der Batterie und die anderen sind unbelegt? Wenn du 50W als Limit einträgst, sind es dann genau 200W auf die der Inverter geht (Bisher 200W * 4 channels = 800W)? Dann muss ich mir mal die Channelerkennung und Kalkulation anschauen. Habe nur einen HM-300 mit einem Channel. Da ist das einfach.

**Reply by l3ztum** [reply]
Das könnte sein, ich teste heute Abend mal ob deine Theorie passt. Genau die Batterie hängt aktuell nur an einem Channel, die entsprechenden Y-Splitter sind noch nicht da.

**Reply by l3ztum** [reply]
@helgeerbe 
Ich habe nochmal ein wenig analysiert um dabei sind mir folgende Dinge aufgefallen:

Ich muss für meinen Anwendungsfall wahrscheinlich auf beide MPPT Regler gehen, da mir sonst eine Fehlermeldung mit MPPT-B undervolt im InverterLog auftaucht und ich nicht auf die volle eingestellte Leistung komme.

Du hattest Recht mit dem Output auf 50W. Die Console zeigt an der Stelle dann wie erwartet ein Set auf 50W `20:42:24.217 > [PowerLimiterClass::loop] dcVoltage: 37.70 Voltage Start Threshold: 36.00 Voltage Stop Threshold: 32.00 inverter->isProducing(): 0
20:42:24.217 > [PowerLimiterClass::loop] newPowerLimit: 50
20:42:24.217 > [PowerLimiterClass::loop] Starting up inverter...`. Der Output im LiveView ist dann allerdings auf 200W gesetzt. Ich schau auch mal bei Gelegenheit auf den Code, evtl. fällt es mir auch auf. Wie unten im zu sehen. 
Soll ich dafür ein Bug-Issue aufmachen?

Hier noch ein kleiner Nitpick. Es gibt aktuell keine Möglichkeit den dynamischen Power Limiter im LiveView zu überschreiben. Möglichkeit via MQTT ist gegeben aber in der Oberfläche muss man erst ins Settings-Untermenü. Wenn man wie ich panisch im LiveView auf Inverter ausschalten drückt, weil die Ströme viel zu hoch sind wird dieser munter wieder vom Dynamic Power Limiter angeschaltet. Ist in dem Sinne kein Fehler aber man muss erstmal drauf kommen. 

Ich schaue die Tage mal wie ich bei den Punkten unterstützen kann, wenn ich Zeit finde mach ich evtl. mal einen PR auf. Finde es auf jeden Fall super alles auf einem ESP zu haben abseits der Heimautomation. 👍

![Bild](https://user-images.githubusercontent.com/26739228/235371352-f316dd99-2dae-4a13-b1c6-4e5b7da16459.png)

**Reply by helgeerbe** [reply]
Code ist from Pr #156. I don't have a Multichannel device, and don't know how it behave. I would appreciates if you can have a look at the code.

**Reply by peff74** [reply]
Hi,
Also bei meinem HM700 mit zwei Kanälen hängt z.Z. auch nur der Akku an einem Kanal.
Regelung funktioniert einwandfrei, jetzt mit der neuen Version.

Regelung könnte ggf. etwas träger oder deutlich schneller sein. Wenn z.B. ein grosser Verbraucher pulst (Ceranfeld ) läuft die Regelung hinterher. Also entweder schneller was wahrscheinlich nicht geht, oder träger damit es ignoriert wird.

**Reply by l3ztum** [reply]
Ich konnte noch keine weitere Zeit reinstecken aber im genannten PR wird Dynamic Limiter Output mit Anzahl der aktiven Channel multipliziert. Der HM verhält sich aber eher als ob die Anzahl der MPPT-Regler mit dem Setpoint des Dynamic Limiter multipliziert werden sollte. Das würde am Verhalten z. B. deines HM-700 nichts ändern aber der HM1500 würde nicht übersteuern.

**Reply by l3ztum** [reply]
@MalteSchm you also noticed problems with the hm 1500 and the dynamic power Limiter based on active dc inputs in the mentioned pr #156.

My guess is that the power value should be multiplied by the count of mppt regulators instead by dc outputs. 
I looked through the code but unfortunately I did not find any solution to check for available mppt regulators via the inverter infos. @helgeerbe did I miss something or is this currently not served via the dtu interface? This would then require some lookup table and manually manged data for the supported controllers. Any input would be Highly appreciated.

**Reply by helgeerbe** [reply]
@l3ztum I don't know if you can query the number of MPPTs.

**Reply by l3ztum** [reply]
A comment on a issue in opendtu seems to confirm my hypothesis that the hm-1500 splits 50/50 between channel 1/2 and 3/4. https://github.com/tbnobody/OpenDTU/issues/272#issuecomment-1292102738

**Reply by MalteSchm** [reply]
Yes it seems that the a limit of e.g. 50% sets the limits of each input converter (i.e. MPPT tracker) to 50%. So if you only have on input channel connected the current calculation in the Power Limiter (which uses active channels) is wrong.
You can do the math if you take e.g. test 102 from the attached xls sheet. 

This plus the issue that you will cross the 11.5A point (at which the HM 1500 stops to regulate the limit properly) has lead me to the believe: Don't do this. I have at least decided against this. 

[Messungen_Hoymiles.xlsx](https://github.com/helgeerbe/OpenDTU-OnBattery/files/11380335/Messungen_Hoymiles.xlsx)

**Comment by bf4ed** [comment]
Es wäre auch gut jede Phase einzeln auf Live view zu sehen. Als auswahl zwischen total grid power und L1,L2,L3
